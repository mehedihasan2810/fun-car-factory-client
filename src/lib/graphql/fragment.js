import gql from "graphql-tag";

export const CORE_CARS_FIELD = gql`
  fragment CoreCarsField on Car {
    id
    name
    price
    url
    rating
    category
    quantity
  }
`;
