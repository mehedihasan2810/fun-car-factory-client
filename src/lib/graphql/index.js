import {
  ApolloClient,
  InMemoryCache,
  createHttpLink,
  from,
} from "@apollo/client";
import { setContext } from "@apollo/client/link/context";
import { onError } from "@apollo/client/link/error";
import { signOut } from "firebase/auth";
import Cookies from "js-cookie";
import { auth } from "../../configs/firebase";

// Create an in-memory cache
export const cache = new InMemoryCache();

// Create an HTTP link to the GraphQL server
const httpLink = createHttpLink({
  uri: "https://fun-car-factory-server.vercel.app/graphql",
  // uri: "http://localhost:4000/graphql",
});

// Set the authorization context based on the user's token
const authLink = setContext((_, { headers }) => {
  const token = Cookies.get("token");

  return {
    headers: {
      ...headers,
      authorization: token ? `Bearer ${token}` : "",
    },
  };
});

// Handle errors, including authentication errors
const errorLink = onError(async ({ graphQLErrors, networkError }) => {
  if (graphQLErrors) {
    if (graphQLErrors[0]?.message === "User is not authenticated") {
      // Sign out the user, remove the token, and clear the Apollo Client store
      await signOut(auth);

      Cookies.remove("token");

      apolloClient.clearStore();
    }

    // Log GraphQL errors
    graphQLErrors.forEach(({ message }) =>
      console.error(`[GraphQL error]: Message: ${message}`)
    );
  }
  if (networkError) console.error(`[Network error]: ${networkError}`);
});

// Create the Apollo Client with the configured link and cache
export const apolloClient = new ApolloClient({
  // link: authLink.concat(errorLink, httpLink),
  link: from([authLink, errorLink, httpLink]),
  cache,
});
