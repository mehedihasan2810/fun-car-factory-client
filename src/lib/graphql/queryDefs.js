import { gql } from "@apollo/client";

export const GET_CARS = gql`
  query GetCars {
    getCars {
      id
      name
      price
      url
      rating
    }
  }
`;

export const GET_CAR = gql`
  query GetCar($id: String!) {
    getCar(id: $id) {
      id
      name
      price
      rating
      url
      sellerName
      email
      category
      description
    }
  }
`;
