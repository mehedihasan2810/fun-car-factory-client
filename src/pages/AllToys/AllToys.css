.all-toys-container {
  margin: 100px 0;
}
.all-toys-container table {
  border-collapse: collapse;
  border-spacing: 0;
  width: 100%;
  border: 1px solid #ddd;
}

.all-toys-container th {
  background: rgba(0, 0, 255, 0.112);
}

.all-toys-container th,
td {
  text-align: left;
  padding: 16px;
}

.all-toys-container tr:nth-child(even) {
  background-color: #f2f2f2;
}

.all-toys-container img {
  width: 35px;
  height: 35px;
  object-fit: cover;
  border-radius: 9999px;
}

.all-toys-container .name-td {
  display: flex;
  align-items: center;
  gap: 3px;
}
