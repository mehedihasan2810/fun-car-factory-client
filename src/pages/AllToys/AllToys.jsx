import { useLoaderData } from "react-router-dom";
import "./AllToys.css";
const AllToys = () => {
  const allToys = useLoaderData();
  console.log(allToys);

  return (
    <div className="center-container">
      <div className="all-toys-container">
        <table>
          <thead>
            <tr>
              <th>Seller Name</th>
              <th>Toy Name</th>
              <th>Sub Category</th>
              <th>Price</th>
              <th>Quantity</th>
              <th>View Details</th>
            </tr>
          </thead>
          <tbody>
            {allToys.map((toy) => (
              <tr key={toy._id}>
                <td>{toy.sellerName}</td>
                <td className="name-td">
                  <img src={toy.url} alt="" />
                  {toy.name}
                </td>
                <td>{toy.category}</td>
                <td>{toy.price}$</td>
                <td>{toy.quantity}</td>
                <td>
                  <button className="btn-primary">View Details</button>
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    </div>
  );
};

export default AllToys;
