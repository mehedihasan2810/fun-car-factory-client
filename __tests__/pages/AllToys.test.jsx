import { ApolloProvider } from "@apollo/client";
import {
  customRender,
  describe,
  expect,
  test,
} from "../utils/utils";
import { apolloClient } from "../../src/lib/graphql";
import AllToys from "../../src/pages/AllToys/AllToys";

describe("<AllToys />", () => {
  test("Should render successfully", async () => {
    // ARRANGE
    customRender(
      <ApolloProvider client={apolloClient}>
        <AllToys />
      </ApolloProvider>
    );
  });

  test("Should fetch all cars on render and match snapshot", async () => {
    // ARRANGE
    const { asFragment } = customRender(
      <ApolloProvider client={apolloClient}>
        <AllToys />
      </ApolloProvider>
    );

    // ASSERT
    expect(asFragment()).toMatchSnapshot();
  });
});
