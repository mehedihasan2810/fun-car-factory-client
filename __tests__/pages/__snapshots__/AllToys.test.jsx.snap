// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<AllToys /> > Should fetch all cars on render and match snapshot 1`] = `
<DocumentFragment>
  <section
    class="toys-container"
  >
    <div
      class="toys-top-header"
    >
      <div
        class="toys-total-count"
      >
        Toys 
        <span>
          (0)
        </span>
      </div>
      <div
        class="search-form-control"
      >
        <svg
          class="w-6 h-6"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21 21l-5.197-5.197m0 0A7.5 7.5 0 105.196 5.196a7.5 7.5 0 0010.607 10.607z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
        <input
          data-testid="search-input"
          id="search"
          name="search"
          placeholder="Search Your Favorite Toys... eg. bus, ferrari, truck"
          type="text"
          value=""
        />
        <button
          data-testid="search-btn"
          type="submit"
        >
          Search
        </button>
      </div>
      <div
        class="toys-filter-sortby-wrapper"
      >
        <div
          class="toys-filter-wrapper"
        >
          <button
            data-testid="all-toys-filter-btn"
          >
            <span>
              Filter
            </span>
            <span>
              <svg
                class="w-6 h-6"
                fill="none"
                stroke="currentColor"
                stroke-width="1.5"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10.5 6h9.75M10.5 6a1.5 1.5 0 11-3 0m3 0a1.5 1.5 0 10-3 0M3.75 6H7.5m3 12h9.75m-9.75 0a1.5 1.5 0 01-3 0m3 0a1.5 1.5 0 00-3 0m-3.75 0H7.5m9-6h3.75m-3.75 0a1.5 1.5 0 01-3 0m3 0a1.5 1.5 0 00-3 0m-9.75 0h9.75"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </span>
          </button>
          <div
            class="toys-filter-options"
            data-testid="all-toys-filter-options"
          >
            <button>
              All
            </button>
            <button>
              Ferrari
            </button>
            <button>
              Bus
            </button>
            <button>
              Truck
            </button>
          </div>
        </div>
        <div
          class="toys-sortby-wrapper"
        >
          <button
            data-testid="all-toys-sortby-btn"
          >
            <span>
              Sort By
            </span>
            <span>
              <svg
                class="w-6 h-6"
                fill="none"
                stroke="currentColor"
                stroke-width="1.5"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19.5 8.25l-7.5 7.5-7.5-7.5"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
              <svg
                class="toys-sort-upper-chevron"
                fill="none"
                stroke="currentColor"
                stroke-width="1.5"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.5 15.75l7.5-7.5 7.5 7.5"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </span>
          </button>
          <div
            class="toys-sortby-options"
            data-testid="all-toys-sortby-options"
          >
            <button>
              Default
            </button>
            <button>
              Price: Low To High
            </button>
            <button>
              Price: High To Low
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="toys-grid"
    >
      <div
        style="display: none;"
      >
        all-toys-loading
      </div>
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
          style="height: 400px; border-radius: 15px;"
        >
          ‌
        </span>
        <br />
      </span>
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
          style="height: 400px; border-radius: 15px;"
        >
          ‌
        </span>
        <br />
      </span>
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
          style="height: 400px; border-radius: 15px;"
        >
          ‌
        </span>
        <br />
      </span>
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
          style="height: 400px; border-radius: 15px;"
        >
          ‌
        </span>
        <br />
      </span>
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
          style="height: 400px; border-radius: 15px;"
        >
          ‌
        </span>
        <br />
      </span>
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
          style="height: 400px; border-radius: 15px;"
        >
          ‌
        </span>
        <br />
      </span>
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
          style="height: 400px; border-radius: 15px;"
        >
          ‌
        </span>
        <br />
      </span>
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
          style="height: 400px; border-radius: 15px;"
        >
          ‌
        </span>
        <br />
      </span>
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
          style="height: 400px; border-radius: 15px;"
        >
          ‌
        </span>
        <br />
      </span>
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
          style="height: 400px; border-radius: 15px;"
        >
          ‌
        </span>
        <br />
      </span>
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
          style="height: 400px; border-radius: 15px;"
        >
          ‌
        </span>
        <br />
      </span>
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
          style="height: 400px; border-radius: 15px;"
        >
          ‌
        </span>
        <br />
      </span>
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
          style="height: 400px; border-radius: 15px;"
        >
          ‌
        </span>
        <br />
      </span>
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
          style="height: 400px; border-radius: 15px;"
        >
          ‌
        </span>
        <br />
      </span>
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
          style="height: 400px; border-radius: 15px;"
        >
          ‌
        </span>
        <br />
      </span>
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
          style="height: 400px; border-radius: 15px;"
        >
          ‌
        </span>
        <br />
      </span>
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
          style="height: 400px; border-radius: 15px;"
        >
          ‌
        </span>
        <br />
      </span>
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
          style="height: 400px; border-radius: 15px;"
        >
          ‌
        </span>
        <br />
      </span>
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
          style="height: 400px; border-radius: 15px;"
        >
          ‌
        </span>
        <br />
      </span>
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
          style="height: 400px; border-radius: 15px;"
        >
          ‌
        </span>
        <br />
      </span>
    </div>
  </section>
</DocumentFragment>
`;
